### PR TITLE
Implemented add Opportunity/Fix Mapping

### PR DESCRIPTION
**Pull request by amDavid01**
https://app.shortcut.com/deskpro/story/75729/salesforce-add-opportunity
https://app.shortcut.com/deskpro/story/98469/salesforce-fix-mapping
This also fixed the mapping issue

### CODE REVIEW

File: src/pages/create/Opportunity.tsx [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  const [privateParam, stage, type, description, dlStatus] = watch([
+    "Private",
+    "StageName",
+    "LeadSource",
+    "Description",
+    "DeliveryInstallationStatus__c",
+  ]);
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const submit = async (data: any) => {
```
Comment (zpawn) [inline src/pages/create/Opportunity.tsx:111]: Why any? Let's typify submitting data

File: src/pages/create/Opportunity.tsx [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            }}
+          />
+        </Stack>
+        <Stack
+          style={{ justifyContent: "space-between", width: "100%" }}
+          gap={5}
+        >
+          <Button
+            type="submit"
+            text={submitting ? "Creating..." : "Create"}
```
Comment (zpawn) [inline src/pages/create/Opportunity.tsx:173]: In this case, we usually `disabled` button and show the `loading` icon

here example https://github.com/DeskproApps/gitlab/blob/master/src/components/IssueForm/IssueForm.tsx#L127

File: src/pages/create/Opportunity.tsx (hunk 1 above)
Comment (amDavid01) [reply src/pages/create/Opportunity.tsx:111]: Can't, the type comes from the metadata, I did this on activities and ash said i don't need it

**Review by zpawn — APPROVED** [review]

File: src/pages/create/Opportunity.tsx [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    [QueryKey.OPPORTUNITY_METADATA, "Opportunity"],
+    (client) => getObjectMeta(client, "opportunity")
+  );
+
+  useEffect(() => {
+    if (!opportunityMetadata.data) return;
+
+    setSchema(
+      getMetadataBasedSchema(opportunityMetadata.data.fields, {
+        Name: z.string().min(1),
```
Comment (AshleyDawson) [inline src/pages/create/Opportunity.tsx]: We always need to build the validation schema from sObject metadata as we can't guarantee these fields will exist or be mandatory.

There shouldn't be any circumstance where there are hard coded sObject fields.

Comment (AshleyDawson) [inline src/pages/create/Opportunity.tsx:75]: We always need to build the validation schema from sObject metadata as we can't guarantee these fields will exist or be mandatory.

There shouldn't be any circumstance where there are hard coded sObject fields.

File: src/schemas/default.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+  for (const key of Object.keys(customInputs)) {
+    newObj[key as keyof typeof newObj] =
+      customInputs[key as keyof typeof newObj];
+  }
+
+  const schema = z
+    .object({
+      ...newObj,
+      WhoId: z.string().min(1).optional(),
```
Comment (AshleyDawson) [inline src/schemas/default.ts]: We always need to build the validation schema from sObject metadata as we can't guarantee these fields will exist or be mandatory.

There shouldn't be any circumstance where there are hard coded sObject fields.

(as above)

**Review by AshleyDawson — CHANGES_REQUESTED** [review]

File: src/pages/create/Opportunity.tsx [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+  });
+
+  useEffect(() => {
+    if (parentId) {
+      setValue(object as string, parentId);
+    }
+  }, [parentId, setValue, object]);
+
+  useEffect(() => {
+    register("StageName", {
```
Comment (AshleyDawson) [inline src/pages/create/Opportunity.tsx]: What is the purpose of these hard-coded field names? `StageName` and `CloseDate`

**Pull request merged**